en/lessons/libraries/bypass.md

```diff
@@ -0,0 +1,274 @@
+  end
+end
+```
+
+Our test is simple enough and if we run it we'll see it passes but let's dig in and see what each portion is doing.
+The first thing we see in our test is the `Bypass.expect/2` function:
+
+```elixir
+Bypass.expect(bypass, fn conn ->
+  Plug.Conn.resp(conn, 200, "pong")
+end)
+
+```
+
+`Bypass.expect/2` takes our Bypass connection and a single arity function which is expected to modify a connection and return it, this is also an opportunity to make assertions on the request to verify it's as we expect.  Let's update our test url to include `/ping` and assert both the request path and HTTP method:
+
+```elixir
+test "request with HTTP 200 response", %{bypass: bypass} do
+  Bypass.expect(bypass, fn conn ->
+    assert "GET" == conn.method
+    assert "/ping" == conn.request_path
+    Plug.Conn.resp(conn, 200, "pong")
+  end)
+
+  assert {:ok, "pong"} = HealthCheck.ping("http://localhost:#{bypass.port}/ping")
+end
+```
+
+The last part of our test we use `HealthCheck.ping/1` and assert the response is as expected, but what's `bypass.port` all about?
+Bypass is actually listening to a local port and intercepting those requests, we're using `bypass.port` to retrieve the default port since we didn't provide on in `Bypass.open/1`.
+
+Next up is adding test cases for errors.
+We can start with a test much like our first with some minor changes: returning 500 as the status code and assert the `{:error, reason}` tuple is returned:
+
+```elixir
+test "request with HTTP 500 response", %{bypass: bypass} do
+  Bypass.expect(bypass, fn conn ->
+    Plug.Conn.resp(conn, 500, "Server Error")
+  end)
+
+  assert {:error, "HTTP Status 500"} = HealthCheck.ping("http://localhost:#{bypass.port}")
+end
+```
+
+There's nothing special to this test case so let's move on to the next: unexpected server outages, these are the request we're most concerned with.
+To accomplish this we won't be using `Bypass.expect/2` instead we're going to rely on `Bypass.down/1` to shut down the connection:
+
+```elixir
+test "request with unexpected outage", %{bypass: bypass} do
+  Bypass.down(bypass)
+
+  assert {:error, :econnrefused} = HealthCheck.ping("http://localhost:#{bypass.port}")
+end
+```
+
+If we run our new tests we'll see everything passes as expected!
+With our `HealthCheck` module tested we can move on to testing it together with our GenServer based-scheduler.
+
+## Multiple external hosts
+
+For our project we'll keep the scheduler barebones and rely on `Process.send_after/3` to power our reoccuring checks, for more on the `Process` module take a look at the [documentation](https://hexdocs.pm/elixir/Process.html).
+Our scheduler requires three options: the collection of sites, the interval of our checks, and the module the implements `ping/1`.
+By passing in our module we further decouple our functionality and our GenServer, enabling us to better test each in isolation:
+
+```elixir
+defmodule Clinic.Scheduler do
+  use GenServer
+
+  require Logger
```

Comment:
As a personal question that will help me write better code, why do you require Logger while below you are explicitely stating `Logger.info` and `Logger.error`? Is it a good practice?

Reply:
@gemantzu Logger must be required to use it in Elixir, if you remove the `require` and run the code this is the error you should see:

```elixir
you must require Logger before invoking the macro Logger.info/1
```

Reply:
Ok, thnx.